README
ExaBGP is a BGP engine (RFC4271)

It can :

 - announce route as defined by it JunOS looking configuration file
 - generate BGP updates from third party applications
 - parse BGP updates and send their textual representations to thrid party applications

Potential use are :

 - announce IPv4/IPv6 routes with arbitrary next-hops into your network
   * injection of service IPs like AS112 announcement
   * temporary route redirection (adding more specific routes with different next-hop)
   * injection of flow routes to handle DDOS
 - source flow routes.
 - track changes in the global routing table.
 - add the received default route to an host
 - it can connect using either IPv4 or IPv6 and announce both IPv4 and IPv6 routes.

Keep up to date, follow us on twitter, user exabgp ( https://twitter.com/#!/search/exabgp )

The program does not perform any FIB manipulation, however it can call an application which will perform them.

The code is deployed in many networks but as noone is perfect,
I will gladly fix bugs if any are found, just contact me: thomas . mangin AT exa - networks . co . uk

As the program is not intended to receive routing information and perform FIB manipulation it only partially implements:
 - RFC 1997 - BGP Communities Attribute
 - RFC 2545 - Use of BGP-4 Multiprotocol Extensions for IPv6 Inter-Domain Routing
 - RFC 4724 - Graceful Restart Mechanism for BGP
 - RFC 4760 - Multiprotocol Extensions for BGP-4, Obsoletes: 2858
 - RFC 4271 - A Border Gateway Protocol 4 (BGP-4), Obsoletes: 1771
 - RFC 5492 - Capabilities Advertisement with BGP-4, Obsoletes 3392,2842
 - RFC 4360 - BGP Extended Communities Attribute
 - RFC 5575 - Dissemination of Flow Specification Rules
 - RFC 4893 - BGP Support for Four-octet AS Number Space
 - RFC 6286 - Autonomous-System-Wide Unique BGP Identifier for BGP-4

The following BG4 related RFC are not supported but would be on the roadmap if there was one :
  * [http://www.ietf.org/rfc/rfc3107.txt RFC 3107] (Carrying Label Information in BGP-4)
  * [http://tools.ietf.org/html/draft-raszuk-idr-flow-spec-v6-01 draft-raszuk-idr-flow-spec-v6-01] (Flow Spec for IPv6)
  * Perhaps [http://tools.ietf.org/html/draft-wkumari-idr-as0-01 draft-wkumari-idr-as0-01] (Codification of AS 0 processing)

Some speakers may use RFC 2385 (Protection of BGP Sessions via the TCP MD5 Signature) to allow the BGP connection to be established.
This is currently only supported under Linux. (The Python socket module exports IPPROTO_TCP but not TCP_MD5_AUTH so it has to be done "by hand").

The following BG4 related RFC are not supported but would be on the roadmap if there was one:
 - RFC 3107 (Carrying Label Information in BGP-4)
Adding the feature should be trivial.

There is no support for:
  - RFC 2796 (BGP Route Reflection - An Alternative to Full Mesh IB).
  - the ORIGINATOR_ID and CLUSTER_LIST attributes.
However, as the program does not re-announce routes (unless you make it do so), nothing prevent its use with a Route Reflector.

I have no interrest in implementing:
 - RFC 2918 (Route Refresh Capability for BGP-4)
 - RFC 3065 (Autonomous System Confederations for BGP)
 - RFC 5549 (Advertising IPv4 Network Layer Reachability Information with an IPv6 Next Hop)

Development is done on python 2.7, the code was made compatible with python 2.4, previous versions can not be supported due to their lack of support for coroutine (the yield feature).

This programs does not have any dependences on any third party libraries.

The documentation is at http://code.google.com/p/exabgp/w/list


CHANGELOG
Version explained:
 major : codebase increase on incompatible changes
 minor : increase on risk of code breakage during a major release
 bug   : increase on bug or incremental changes

Version 3.4.28
 * Fix: add support for ipv6 mpls
   patch by: adrian62

Version 3.4.27
 * Fix: Send state down messages on peer removal and shutdown
 * Improvement: RFC 7674 - Support redirect traffic to a VRF encoded as 4-octet AS, 2-octet Value
   patch by: Omri Matitiau
 * Fix: correctly decode flow routes with multiple NLRI
   reported by: adrian62
 * Improvement: exit with error code 0 on SIGTERM
   reported by: Johan Guldmyr
 * Fix: issue with IPv6 nlri-mpls update generation
   reported by: adrian62

Version 3.4.26
 * Add: JSON notification messages now include the 'neighbor' structure
    that contains the details about which peer the notify was sent from.

Version 3.4.25
  * Fix: support range syntax in Flows protocol, added protocol validation
    patch by: Omri Matitiau

Version 3.4.24
  * Fix: default for IPv6 should be 128 .. not 32 !

Version 3.4.23
  * Fix: issue parsing some flowspec routes (backport from master)

Version 3.4.22
  * Fix: issue with very fast route flapping
    patch by: Ravikumar Arunachalam

Version 3.4.21
  * Fix: make healthcheck work on netbsd
    patch by: he32
  * Fix: re-fix issue #493 Multiple IPv6 MPREACH attributes in one UPDATE 
  * Fix: issue #732 --test not working for eBGP setup
  * Fix: possibly missing API message when sent at high rate, issue #736
    patch by: Ravikumar Arunachalam

Version 3.4.20
  * Fix: profiling
  * Fix: potential issue with networking (select on closed FD/TCP session)
  * Fix: issue with route replay order when sent fast
    patch by: Ravikumar Arunachalam

Version 3.4.19
 * Add: IPv6 nlri-mpls to list of enabled protocol (was missing)
    requested by: adrian62
 * Fix: encoding of Flow Label requiring more than 2 bytes
    reported by: BLAKEMMM
 * Fix: decoding of capability (was potentially over reading)
 * Fix: trace when trying to access PID file and this is not allowed
    reported by: George Shuklin
 * Fix: Remove a peer's RIB cache when it is deleted from the config file
    patch by: Brian Johnson
 * Fix: do not crash the reactor when an invalid IP is passed via the API
    reported by: Yevgeniy Ovsyannikov
 * Fix: bad defintion of Flow for ICMPType, ICMPCode and Fragment
    reported by: Christoph Loibl
 * Feature: allow add-path for mpls-vpn
    reported by: adrian62
 * Change: Backported setup.py from master
 * Feature: added SRPMS for exabgp
    patch by: Leonardo Amaral

Version 3.4.18
 * Backport: backhole community (RFC 7999)
    original patch by: Job Snijders
 * Fix: Configuration parser does not accept configs without neighbors.
    patch by doddt
 * Fix: 'connect' keyword is now also allowed in neighbor scope
    patch by: Stacey Sheldon (Corsa)
 * Fix: removing protocol auto-cleanup (it should never be called and seems to cause a CG issue)
    reported by: Colin Petrie
 * Change: default to a 0 offset for ipv6 flowspec source/destination match
    patch by: Brian Johnson
 * Fix: Better PID file handling
    reported by: Ben Agricola
 * Fix: Update RIB cache families on configuration reload
    patch by: Brian Johnson
 * Fix: Backport fix on SIGUSR2 (restarting process not needing to be)
    patch by: Shawn Zhou
 * Change: group-updates now generates one UPDATE per address family (and not one per NLRI for non IPv4)
    patch by: Brian Johnson

Version 3.4.17
 * Fix: does not accept IPv6 as router-id
    reported by: yuriya
 * Fix: JSON output for flow routes with rd
    reported by droon5
 * Fix: Fix Path-Information
 * Fix: Bad encoding of capability when multiple families are used for add-path
    reported: by Alexander Bespalov
 * Fix: support non ASN4 use of AS_TRANS (AS23456)
    reported by: Todd Crane
 * Fix: do not exit when we can not accept incoming connection
    reported by: Pavel Batkov
 * Fix: quote where not escaped in JSON reason field
    reported by: Rob Barnes
 * Fix: decoding of IPv6 flow routes
    reported by: stoffi92
 * Fix: decoding of Graceful Restart Capability
    patch by: florinz
 * Fix: ASN4 encoding
    patch by: Shu Sugimoto and Eiichiro Watanabe
 * Change: Run without even peers configured
    patch by: Jordan Gedney
 * Fix: JSON encoding of updates without NLRIs
    patch by: Dhammika Pathirana
 * Fix: Possible race conditions in api handling
    patch by: Brian Johnson
 * Feature: Add 'show neighbor status' api
    patch by: Brian Johnson
 * Fix: flush route api
    patch by: Brian Johnson
 * Fix: Allow asn4 peer to speak with asn2 only peer
    patch by: Brian Johnson
 * Fix: only one MP NLRI is allow per UPDATE
    reported by: subsecond
 * Change: configuration output does not includes ':' anymore
    patch by: doddt
 * Change: syslog format changed to be in line with other application
    patch by: Brian Johnson

Version 3.4.16
 * Feature: allow users to decide if processes must be run before or after we drop privileges
    requested by: Ben Agricola
 * Fix: correctly look in /etc/exabgp for programs to run when the path is relative
    reported by: Vincent Bernat
 * Fix: missing handler for NOTIFICATION
    patch by: minglvyy

Version 3.4.15
 * Fix: the ttl-security parameter didn´t really work. Fixed for outgoing connections now.
    patch by: Borja Marcos
 * Fix: configuration leak between processes for neighbor-changes and send-packets
    reported by: spakka
 * Feature: add per neighbor connection port
    requested by: dbarrosop
 * Fix: ASN4 boundary off by one
 * Fix: Bad peer IP when using show routes
    patch by (backported): Wayne Tucker
 * Fix: Missing next-hop in the text api
    reported by: Lisa Roach
 * Fix: broken route-refresh command
    reported by: Bryan Schwerer
 * Fix: wrongly announcing connection issue with peer on the API
    reported by: Bryan Schwerer

Version 3.4.14
  * Change: This version does not exists

lib/bgp/data.py
#!/usr/bin/env python
# encoding: utf-8
"""
data.py

Created by Thomas Mangin on 2009-08-25.
Copyright (c) 2009 Exa Networks. All rights reserved.
"""

HOLD_TIME=180

import math
import time
import socket
from struct import pack,unpack
	# !L : Network Long  (4 bytes)
	# !H : Network Short (2 bytes)

CAFE = chr(202) + chr(254)

# http://www.iana.org/assignments/address-family-numbers/
class AFI (int):
	ipv4 = 0x01
	ipv6 = 0x02

	def __str__ (self):
		if self == 0x01: return "IPv4"
		if self == 0x02: return "IPv6"
		return "unknown afi"

	def pack (self):
		return pack('!H',self)

# http://www.iana.org/assignments/safi-namespace
class SAFI (int):
	unicast = 1					# [RFC4760]
	multicast = 2				# [RFC4760]
#	deprecated = 3				# [RFC4760]
#	nlri_mpls = 4				# [RFC3107]
#	mcast_vpn = 5				# [draft-ietf-l3vpn-2547bis-mcast-bgp] (TEMPORARY - Expires 2008-06-19)
#	pseudowire = 6				# [draft-ietf-pwe3-dynamic-ms-pw] (TEMPORARY - Expires 2008-08-23) Dynamic Placement of Multi-Segment Pseudowires
#	encapsulation = 7			# [RFC5512]
#
#	tunel = 64					# [Nalawade]
#	vpls = 65					# [RFC4761]
#	bgp_mdt = 66				# [Nalawade]
#	bgp_4over6 = 67				# [Cui]
#	bgp_6over4 = 67				# [Cui]
#	vpn_adi = 69				# [RFC-ietf-l1vpn-bgp-auto-discovery-05.txt]
#
#	mpls_vpn = 128				# [RFC4364]
#	mcast_bgp_mpls_vpn = 129	# [RFC2547]
#	rt = 132					# [RFC4684]
#	flow_ipv4 = 133				# [RFC5575]
#	flow_vpnv4 = 134			# [RFC5575]
#
#	vpn_ad = 140				# [draft-ietf-l3vpn-bgpvpn-auto]
#
#	private = [_ for _ in range(241,254)]	# [RFC4760]
#	unassigned = [_ for _ in range(8,64)] + [_ for _ in range(70,128)]
#	reverved = [0,3] + [130,131] + [_ for _ in range(135,140)] + [_ for _ in range(141,241)] + [255,]	# [RFC4760]

	def __str__ (self):
		if self == 0x01: return "unicast"
		if self == 0x02: return "multicast"
		return "unknown safi"

	def pack (self):
		return chr(self)

class IP (object):
	def __init__ (self,value):
		v4 = False
		try:
			pack = socket.inet_pton(socket.AF_INET, str(value))
			numeric = unpack('>L',pack)[0]
			string = str(value)
			v4 = True
		except socket.error:
			pass
		except TypeError:
			pass

		if not v4:
			try:
				pack = socket.inet_pton(socket.AF_INET6, str(value))
				a,b,c,d = unpack('>LLLL',pack)
				numeric = (a << 96) + (b << 64) + (c << 32) + d
				string = str(value).lower()
			except socket.error:
				raise ValueError('"%s" is an invalid address' % str(value))
			except TypeError:
				raise ValueError('"%s" is an invalid address' % str(value))

		self.numeric = numeric
		self._pack = pack
		self.string = string

		if v4:
			self.version = 4
			self.length = 32
		else:
			self.version = 6
			self.length = 128

	def pack (self):
		return self._pack

	def __str__ (self):
		return self.string

	def __len__ (self):
		return self.version

# Super sub-optimal as code... make length as parameter of __init__ ?
class Mask (int):
	def new (self,mask,length):
		#slash_to_size = dict([(length-bits,(1<<bits)) for bits in range(0,length+1)])
		#slash_to_mask = dict([(length-bits,(1L<<length) - (1<<bits)) for bits in range(0,length+1)])
		mask_to_slash = dict([((1L<<length) - (1<<bits),length-bits) for bits in range(0,length+1)])
		ipv4_to_slash = dict([("%d.%d.%d.%d" % (k>>24, k>>16 & 0xFF ,k>>8 & 0xFF, k&0xFF),mask_to_slash[k]) for k in mask_to_slash.keys()])

		#regex_mask = '(?:\d|[12]\d|3[012])'
		try:
			slash = int(mask)
		except ValueError:
			try:
				slash = self.ipv4_to_slash[mask]
			except KeyError:
				raise ValueError('the netmask is invalid %s' % str(mask))

		if not slash in range(0,length+1):
			return ValueError('the netmask is invalid /%s' % str(slash))

		return Mask(slash)

	def pack (self):
		return chr(self)

	def __len__ (self):
		return 1

class RouterID (IP):
	pass

class Version (int):
	def pack (self):
		return chr(self)

class ASN (int):
	# regex = "(?:0[xX][0-9a-fA-F]{1,8}|\d+:\d+|\d+)"
	length = 2

	def four (self):
		self.length = 4
		return self
	
	def pack (self):
		if self.length == 2:
			return pack('!H',self)
		return pack('!L',self)

	def __len__ (self):
		return self.length
	
class Community (long):
	def new (self,data):
		try:
			value = int(data)
		except ValueError:
			separator = data.find(':')
			if separator > 0:
				value = (long(data[:separator])<<16) + long(data[separator+1:])
			elif len(data) >=2 and data[1] in 'xX':
				value = long(data,16)
			else:
				value = long(data)
		return Community(value)

	def pack (self):
		return pack('!L',self)

	def __str__ (self):
		return "%d:%d" % (self >> 16, self & 0xFFFF)

	def __len__ (self):
		return 4


class Communities (list):
	_factory = Community()
	
	def append (self,data):
		list.append(self,self._factory.new(data))
		self.sort()

	def pack (self):
		return ''.join([community.pack() for community in self])

	def __str__ (self):
		if len(self) >  1: return '[ %s ]' % ' '.join([str(community) for community in self])
		if len(self) == 1: return str(self[0])
		return ''

class LocalPreference (long):
	def pack (self):
		return pack('!L',self)

	def __len__ (self):
		return 4

class HoldTime (int):
	def pack (self):
		return pack('!H',self)

	def keepalive (self):
		return int(self/3)

	def __len__ (self):
		return 2

class Flag (int):
	EXTENDED_LENGTH = 0x10 # 16
	PARTIAL         = 0x20 # 32
	TRANSITIVE      = 0x40 # 64
	OPTIONAL        = 0x80 # 128

	def __str__ (self):
		r = []
		v = int(self)
		if v & 0x10:
			r.append("EXTENDED_LENGTH")
			v -= 0x10
		if v & 0x20:
			r.append("PARTIAL")
			v -= 0x20
		if v & 0x40:
			r.append("TRANSITIVE")
			v -= 0x40
		if v & 0x80:
			r.append("OPTIONAL")
			v -= 0x80
		if v:
			r.append("UNKNOWN %s" % hex(v))
		return " ".join(r)

class Origin (int):
	IGP        = 0x00
	EGP        = 0x01
	INCOMPLETE = 0x02

	def __str__ (self):
		if self == 0x00: return 'IGP'
		if self == 0x01: return 'EGP'
		if self == 0x02: return 'INCOMPLETE'
		return 'INVALID'

	def pack (self):
		return chr(self)

class ASPath (int):
	AS_SET      = 0x01
	AS_SEQUENCE = 0x02

	def __str__ (self):
		if self == 0x01: return 'AS_SET'
		if self == 0x02: return 'AS_SEQUENCE'
		return 'INVALID'

class Attribute (int):
	# RFC 4271
	ORIGIN           = 0x01
	AS_PATH          = 0x02
	NEXT_HOP         = 0x03
	MULTI_EXIT_DISC  = 0x04
	LOCAL_PREFERENCE = 0x05
	ATOMIC_AGGREGATE = 0x06
	AGGREGATOR       = 0x07
	COMMUNITY        = 0x08
	# RFC 4760
	MP_REACH_NLRI    = 0x0e # 14
	MP_UNREACH_NLRI  = 0x0f # 15

	def __str__ (self):
		if self == 0x01: return "ORIGIN"
		if self == 0x02: return "AS_PATH"
		if self == 0x03: return "NEXT_HOP"
		if self == 0x04: return "MULTI_EXIT_DISC"
		if self == 0x05: return "LOCAL_PREFERENCE"
		if self == 0x06: return "ATOMIC_AGGREGATE"
		if self == 0x07: return "AGGREGATOR"
		if self == 0x08: return "COMMUNITY"
		if self == 0x0e: return "MP_REACH_NLRI"
		if self == 0x0f: return "MP_UNREACH_NLRI"
		return 'UNKNOWN'

class Parameter (int):
	AUTHENTIFICATION_INFORMATION = 0x01 # Depreciated
	CAPABILITIES                 = 0x02

	def __str__ (self):
		if self == 0x01: return "AUTHENTIFICATION INFORMATION"
		if self == 0x02: return "OPTIONAL"
		return 'UNKNOWN'

# http://www.iana.org/assignments/capability-codes/
class Capabilities (dict):
	RESERVED                 = 0x00 # [RFC5492]
	MULTIPROTOCOL_EXTENSIONS = 0x01 # [RFC2858]
	ROUTE_REFRESH            = 0x02 # [RFC2918]
	OUTBOUND_ROUTE_FILTERING = 0x03 # [RFC5291]
	MULTIPLE_ROUTES          = 0x04 # [RFC3107]
	EXTENDED_NEXT_HOP        = 0x05 # [RFC5549]
	#6-63      Unassigned
	GRACEFUL_RESTART         = 0x40 # [RFC4724]
	FOUR_BYTES_ASN           = 0x41 # [RFC4893]
	# 66 Deprecated
	DYNAMIC_CAPABILITY       = 0x43 # [Chen]
	MULTISESSION_BGP         = 0x44 # [Appanna]
	ADD_PATH                 = 0x45 # [draft-ietf-idr-add-paths]
	# 70-127    Unassigned 
	# 128-255   Reserved for Private Use [RFC5492]

	_unassigned = range(70,128)
	_reserved = range(128,256)

	def __str__ (self):
		r = []
		for key in self.keys():
			if key == self.MULTIPROTOCOL_EXTENSIONS:
				r += ['Multiprotocol for ' + ' '.join(["%s %s" % (str(afi),str(safi)) for (afi,safi) in self[key]])]
			elif key == self.ROUTE_REFRESH:
				r += ['Route Refresh']
			elif key == self.GRACEFUL_RESTART:
				r += ['Graceful Restart']
			elif key == self.FOUR_BYTES_ASN:
				r += ['4Bytes AS %d' % self[key]]
			elif key in self._reserved:
				r += ['private use capability %d' % key]
			elif key in self._unassigned:
				r += ['unassigned capability %d' % key]
			else:
				r+= ['unhandled capability %d' % key]
		return ', '.join(r)

	def default (self):
		self[1] = ((AFI(AFI.ipv4),SAFI(SAFI.unicast)),(AFI(AFI.ipv6),SAFI(SAFI.unicast)))
		return self

	def pack (self):
		rs = []
		for k,vs in self.iteritems():
			for v in vs:
				if k == self.MULTIPROTOCOL_EXTENSIONS:
					d = pack('!H',v[0]) + pack('!H',v[1])
					rs.append("%s%s%s" % (chr(k),chr(len(d)),d))
				else:
					rs.append("%s%s%s" % (chr(k),chr(len(v)),v))
		return "".join(["%s%s%s" % (chr(2),chr(len(r)),r) for r in rs])

class Prefix (object):
	_factory = Mask()
	
	# format is (version,address,slash)
	def __init__ (self,address,mask):
		self.ip = IP(address)
		self.mask = self._factory.new(mask,self.ip.length)

	def bgp (self):
		size = int(math.ceil(float(self.mask)/8))
		return "%s%s" % (self.mask.pack(),self.ip.pack()[:size])

	def __str__ (self):
		return '%s/%d' % (self.ip,self.mask)

class Route (Prefix):
	def __init__ (self,ip,slash,next_hop=''):
		Prefix.__init__(self,ip,slash)
		self.next_hop = next_hop if next_hop else ip
		self.local_preference = 100
		self.communities = Communities()

	def get_next_hop (self):
		return self._next_hop
	def set_next_hop (self,ip):
		self._next_hop = IP(ip)
	next_hop = property(get_next_hop,set_next_hop)

	def get_local_preference (self):
		return self._local_preference
	def set_local_preference (self,preference):
		self._local_preference = LocalPreference(preference)
	local_preference = property(get_local_preference,set_local_preference)

	def __cmp__ (self,other):
		return \
			self.ip == other.ip and \
			self.mask == other.mask and \
			self.next_hop == other.next_hop and \
			self.local_preference == other.local_preference and \
			self.communities == other.communities

	def __str__ (self):
		return "%s next-hop %s%s%s" % \
		(
			Prefix.__str__(self),
			self.next_hop,
			" local_preference %d" % self.local_preference if self.local_preference != 100 else '',
			" community %s" % self.communities if self.communities else ''
		)

	def _attribute (self,attr_flag,attr_type,value):
		if attr_flag & Flag.OPTIONAL and not value:
			return ''
		length = len(value)
		if length > 0xFF:
			attr_flag &= Flag.EXTENDED_LENGTH
		if attr_flag & Flag.EXTENDED_LENGTH:
			len_value = pack('!H',length)[0]
		else:
			len_value = chr(length)
		return "%s%s%s%s" % (chr(attr_flag),chr(attr_type),len_value,value)

	def _segment (self,seg_type,values):
		if len(values)>255:
			return self._segment(values[:256]) + self._segment(values[256:])
		return "%s%s%s" % (chr(seg_type),chr(len(values)),''.join([v.pack() for v in values]))

	def pack (self,local_asn,peer_asn,mp_action=''):
		message = ''
		message += self._attribute(Flag.TRANSITIVE,Attribute.ORIGIN,Origin(Origin.IGP).pack())
		message += self._attribute(Flag.TRANSITIVE,Attribute.AS_PATH,'' if local_asn == peer_asn else self._segment(ASPath.AS_SEQUENCE,[local_asn]))
		if local_asn == peer_asn:
			message += self._attribute(Flag.TRANSITIVE,Attribute.LOCAL_PREFERENCE,self.local_preference.pack())
		message += self._attribute(Flag.TRANSITIVE|Flag.OPTIONAL,Attribute.COMMUNITY,''.join([c.pack() for c in self.communities])) if self.communities else ''
		# we do not store or send MED
		if self.ip.version == 4:
			message += self._attribute(Flag.TRANSITIVE,Attribute.NEXT_HOP,self.next_hop.pack())
		if self.ip.version == 6:
			if mp_action == '-':
				attr = AFI(AFI.ipv6).pack() + SAFI(SAFI.unicast).pack() + Prefix.pack(self)
				message += self._attribute(Flag.TRANSITIVE,Attribute.MP_UNREACH_NLRI,attr)
			if mp_action == '+':
				prefix = Prefix.bgp(self)
				next_hop = self.next_hop.pack()
				attr = AFI(AFI.ipv6).pack() + SAFI(SAFI.unicast).pack() + chr(len(next_hop)) + next_hop + chr(0) + prefix
				message += self._attribute(Flag.TRANSITIVE,Attribute.MP_REACH_NLRI,attr)
		return message

# The definition of a neighbor (from reading the configuration)

class Neighbor (object):
	def __init__ (self):
		self.description = ''
		self._router_id = None
		self.local_address = None
		self.peer_address = None
		self.peer_as = None
		self.local_as = None
		self.hold_time = HoldTime(HOLD_TIME)
		self.routes = []

	def missing (self):
		if self.local_address is None: return 'local-address'
		if self.peer_address is None: return 'peer-address'
		if self.local_as is None: return 'local-as'
		if self.peer_as is None: return 'peer-as'
		if self.peer_address.version == 6 and not self._router_id: return 'router-id'
		return ''


	def get_router_id (self):
		return self._router_id if self._router_id else self.local_address
	def set_router_id (self,id):
		self._router_id = id
	router_id = property(get_router_id,set_router_id)

	def __eq__ (self,other):
		return \
			self._router_id == other._router_id and \
			self.local_address == other.local_address and \
			self.local_as == other.local_as and \
			self.peer_address == other.peer_address and \
			self.peer_as == other.peer_as

	def __ne__(self, other):
		return not (self == other)

	def __str__ (self):
		return """\
neighbor %s {
	description "%s";
	router-id %s;
	local-address %s;
	local-as %d;
	peer-as %d;
	static {%s
	}
}""" % (
	self.peer_address,
	self.description,
	self.router_id,
	self.local_address,
	self.local_as,
	self.peer_as,
	'\n\t\t' + '\n\t\t'.join([str(route) for route in self.routes]) if self.routes else ''
)


# Taken from perl Net::IPv6Addr
#	ipv6_patterns = {
#		'preferred' : """\
#			^(?:[a-f0-9]{1,4}:){7}[a-f0-9]{1,4}$
#		""",
#		'compressed' : """\
#			^[a-f0-9]{0,4}::$
#		|	^:(?::[a-f0-9]{1,4}){1,6}$
#		|	^(?:[a-f0-9]{1,4}:){1,6}:$
#		|	^(?:[a-f0-9]{1,4}:)(?::[a-f0-9]{1,4}){1,6}$
#		|	^(?:[a-f0-9]{1,4}:){2}(?::[a-f0-9]{1,4}){1,5}$
#		|	^(?:[a-f0-9]{1,4}:){3}(?::[a-f0-9]{1,4}){1,4}$
#		|	^(?:[a-f0-9]{1,4}:){4}(?::[a-f0-9]{1,4}){1,3}$
#		|	^(?:[a-f0-9]{1,4}:){5}(?::[a-f0-9]{1,4}){1,2}$
#		|	^(?:[a-f0-9]{1,4}:){6}(?::[a-f0-9]{1,4})$
#		""",
#		'ipv4' : """\
#			^(?:0:){5}ffff:(?:\d{1,3}\.){3}\d{1,3}$
#		|	^(?:0:){6}(?:\d{1,3}\.){3}\d{1,3}$
#		""",
#		'ipv4 compressed' : """\
#			^::(?:ffff:)?(?:\d{1,3}\.){3}\d{1,3}$
#		""",
#	}
#
#	ipv4_patterns = '(?:\d{1,2}|1\d{2}|2[0-4]\d|25[0-5])\.(?:\d{1,2}|1\d{2}|2[0-4]\d|25[0-5])\.(?:\d{1,2}|1\d{2}|2[0-4]\d|25[0-5])\.(?:\d{1,2}|1\d{2}|2[0-4]\d|25[0-5])'


